src/main/java/at/ac/tuwien/kr/alpha/common/atoms/ExternalLiteral.java

```diff
@@ -114,35 +119,82 @@ public ExternalLiteral substitute(Substitution substitution) {
-		List<Term> output = getAtom().getOutput();
-		List<Substitution> substitutions = new ArrayList<>();
 		List<Term> substitutes = new ArrayList<>(input.size());
 
+		// evaluate the external atom
 		for (Term t : input) {
 			substitutes.add(t.substitute(partialSubstitution));
 		}
-
 		Set<List<ConstantTerm<?>>> results = getAtom().getInterpretation().evaluate(substitutes);
-
 		if (results == null) {
 			throw new NullPointerException("Predicate " + getPredicate().getName() + " returned null. It must return a Set.");
 		}
 
-		if (results.isEmpty()) {
-			return emptyList();
+		if (this.isNegated()) {
+			return this.checkNegatedLiteralSatisfied(results) ? Collections.singletonList(partialSubstitution) : Collections.emptyList();
+		} else {
+			return this.buildSubstitutionsForOutputs(partialSubstitution, results);
 		}
+	}
 
-		for (List<ConstantTerm<?>> bindings : results) {
-			if (bindings.size() < output.size()) {
-				throw new RuntimeException("Predicate " + getPredicate().getName() + " returned " + bindings.size() + " terms when at least " + output.size() + " were expected.");
+	/**
+	 * Checks whether this negated external literal is satisfied.
+	 * 
+	 * Note that this method must only be called on negated external literals.
+	 * In that case, the literal itself does not bind any output variables,
+	 * i.e. the underlying atom is satisfied iff the output terms obtained by
+	 * evaluating the underlying external atom match the values to which
+	 * the respective output variables are bound. The result of this method assumes
+	 * that the literal is negated, i.e. for the underlying atom AT, it represents
+	 * the truth value !AT.
+	 * Furthermore, this method assumes that the output terms of the underlying atom
+	 * are {@link ConstantTerm}s, i.e. that the grounder's current partial
+	 * substitution has already been applied to them. That assumption is safe since
+	 * in case of a negated external literal, the output variables are always
+	 * non-binding.
+	 * 
+	 * @param outputs The term lists obtained from evaluating the external atom
+	 *                (i.e. calling the java method) encapsulated by this literal
+	 * @return true iff no list in output equals the external atom's output term
+	 *         list as substituted by the grounder, false otherwise
+	 */
+	private boolean checkNegatedLiteralSatisfied(Set<List<ConstantTerm<?>>> outputs) {
+		List<Term> externalAtomOutTerms = this.getAtom().getOutput();
+		boolean outputMatches;
+		for (List<ConstantTerm<?>> output : outputs) {
+			outputMatches = true;
+			for (int i = 0; i < externalAtomOutTerms.size(); i++) {
+				if (!output.get(i).equals(externalAtomOutTerms.get(i))) {
+					outputMatches = false;
+					break;
+				}
 			}
+			if (outputMatches) {
+				// we found one term list where all terms match the ground output terms of the
+				// external atom, therefore the atom is true and the (negative) literal false.
+				return false;
+			}
+		}
```

Comment:
Using labelled continue, the double-loop would look as follows:

```suggestion
		atomTuplesLoop:
		for (List<ConstantTerm<?>> output : outputs) {
			for (int i = 0; i < externalAtomOutTerms.size(); i++) {
				if (!output.get(i).equals(externalAtomOutTerms.get(i))) {
					continue atomTuplesLoop;
				}
			}
			// Found one term list where all terms match the ground output terms of the
			// external atom, therefore the atom is true and the (negative) literal false.
			return false;
		}
```

Reply:
Personally, I don't think the labelled continue improves readability. I know this is a matter of personal taste, but to me it just looks "weird"... So if you're ok with it, I'd rather refrain from using that...

Reply:
I am fine with that. As we discussed, I think the labelled continue is actually easier to read and so I wanted to at least show it. :-)